### ISSUE TITLE
Add contact groups as attendees for an event [$50]

### ISSUE THREAD

**Issue by tcitworld** [nextcloud/calendar#167]
The participant event module should suggest contacts groups **from the contacts app**.

# Related issues
* To add Nextcloud groups (user management, including LDAP) as attendees for an event, see #2955
* To add a Nextcloud circle as attendees for an event, see #2954
* To add a Nextcloud talk group as attendee for an event, see #2956

**Comment by tcitworld** [comment]
Maybe linked to #163

**Comment by georgehrke** [comment]
I'm actually not sure if we can query for contact groups via the contacts php api.

Also, how to show if some users of a group have no e-mail address

**Comment by tcitworld** [comment]
> Also, how to show if some users of a group have no e-mail address

Just don't show them. :-)
I'll look if the groups are available.

**Comment by maksemuz** [comment]
I'd offer the following:
1 - use access groups, especially synced LDAP groups, not groups from contacts. You can parse LDAP groups with standard LDAP tools.
2 - if some participant account has no email, you can just add event to its calendar but do not send invitation by email.

**Comment by pwFoo** [comment]
Should groups expand to members / contacts and add the "person" to the event or just added as group of persons? 

Expand the groups to users / contacts / persons would be flexible (add / remove single persons from event).
Save group to event would add / remove users / contacts if groups modified (maybe not wanted for fix planned events).

**Comment by georgehrke** [comment]
> Expand the groups to users / contacts / persons would be flexible (add / remove single persons from event).

This would be my choice for a first implementation. That should be fairly easy and doesn't require modifications to the Nextcloud server itself.

**Comment by dieter-wilhelm** [comment]
Added 50 € on Bountysource for this indespensible calendar feature.  But for the sake of privacy there must be an option to hide the list of invitees!

**Comment by flozi00** [comment]
> > Expand the groups to users / contacts / persons would be flexible (add / remove single persons from event).
> 
> This would be my choice for a first implementation. That should be fairly easy and doesn't require modifications to the Nextcloud server itself.

Any updates about implementing this feature as temporary solution ?

**Comment by georgehrke** [comment]
Nope, as always: This is a free-software project. Pull-requests are very much appreciated.

**Comment by flozi00** [comment]
Where can I find the relevant code part ?

**Comment by georgehrke** [comment]
To implement this, you would have to alter the ContactsController:
https://github.com/nextcloud/calendar/blob/master/lib/Controller/ContactController.php
Preferably one endpoint that allows the lookup of groups and another endpoint that returns chunks of contacts in this group.

 We have an [InviteesListSearch component](https://github.com/nextcloud/calendar/blob/master/src/components/Editor/Invitees/InviteesListSearch.vue).

In the search method (`findAttendees`), start looking for groups. When the user selects a group, start loading the contacts of that group in the `addAttendee` method. Please load them in chunks of 50 to 100.

**Comment by flozi00** [comment]
My PHP knowledge is too basic for that and never worked with vue before.
I tried it two days without success, can you implement this feature please?

**Comment by jakobroehrl** [comment]
What do you think about adding the talk room participants as attendees?
Similar function as sharing a file to a talk room

**Comment by wiswedel** [comment]
This discussion and the related issues that have been closed for being duplicates takes some wild directions in my opinion.

Some are talking about contact groups (from the Contacts app), some are talking about User groups (from the user management / LDAP). Even Talk rooms have been brought up.
Additionally, Circles should not be forgotten since they're also a way of grouping people plus you can already share a calendar with a circle, but not an event.

tl;dr: We have the request to share with
1. Contact groups (from address books in the Contacts app)
2. User groups (from the user management, including LDAP groups)
3. Circles
4. Talk rooms (i.e. their members)

If you, @tcitworld (as issue initiator) agree with me, would you mind changing the issue title?

**Comment by pierreozoux** [comment]
This is a feature we'd be interested to. And we could even fincance it, if somebody tells me how much he'd need to develop it (my email is in github if you want to contact me and make me an offer, for us it would be about User groups)

@wiswedel, you are right, it is currently too broad. I think there is a initiative in Nc to rationalize this "grouping" thing.

**Comment by micheledimeo** [comment]
> This is a feature we'd be interested to. And we could even fincance it, if somebody tells me how much he'd need to develop it (my email is in github if you want to contact me and make me an offer, for us it would be about User groups)
> 
> @wiswedel, you are right, it is currently too broad. I think there is a initiative in Nc to rationalize this "grouping" thing.

I'm interested in financing it too.

**Comment by julian70400** [comment]
Any news about this ?

**Comment by m4us1ne** [comment]
we are interested in a solution that includes circles. should we post a bounty on this issue or create another issue directly mentioning circles? @micheledimeo and @pierreozoux : bountysource was a good solution for supporting  an issue and worked for us in the past.

**Comment by micheledimeo** [comment]
@m4us1ne , I felt that my offer was too low and then bountysource is at the centre of too much controversy for my liking.
Since this issue was included in the "approved improvements" of calendar, I'm waiting for version 2.2 to use this feature.
In the meantime I'm using a calendar external to NC.

**Comment by tcitworld** [comment]
> tl;dr: We have the request to share with
>     1. Contact groups (from address books in the Contacts app)
>     2. User groups (from the user management, including LDAP groups)
>     3. Circles
>     4. Talk rooms (i.e. their members)

> If you, @tcitworld (as issue initiator) agree with me, would you mind changing the issue title?

Done and splitted the issue into several (see the updated first message).

**Comment by DJaeger** [comment]
Bounty-URL:
https://www.bountysource.com/issues/39038806-add-contact-groups-as-attendees-for-an-event

**Comment by pierreozoux** [comment]
Looks like bountysource is no longer the default.

@tcitworld do you plan to develop this feature?
If yes, you could as well donate to https://framasoft.org/en/ (he is part of this association)

**Comment by julian70400** [comment]
Hello, any news ?

This feature is absoluelty essentiel. How can we help ? What is actually blocking this feature ?

**Comment by jpubb** [comment]
> I'm actually not sure if we can query for contact groups via the contacts php api.
> 
> Also, how to show if some users of a group have no e-mail address

This is what people want to use...

![image](https://user-images.githubusercontent.com/71899277/118851362-8287d300-b886-11eb-9b65-6b73ac0af385.png)

People are simply trying to replicate MS Exchange Distribution Lists functionality here (used for decades).  Nobody want's to have to onesie twosie manually add recips.   The calendar app already ignores lookups/results where the nextcloud user does not have an email field populated.  

Why not just do something like:

**Option 1:**
- Let use select a group(s) as attendee 
- Upon selection, **ValidateGroup()** and  foreach user in group, verify each user in the group has an email address, or throw error that "$CurrentGroupMemberNAme does not have an email address associated with their account. Terminating."

Keep it simple. Admins will know what to do from there (add an email address for the user, or remove the user from the group)

**Option 2:**
- Let use select a group(s) as attendee 
- Upon selection, **ExpandGroup()** and foreach user in group who has an email add that individual attendee in the TO or CC field.


**Option 3**:
Worst option, but I think people will take what they can get here because manually adding people sucks to the point of nothing nice to say.  If you're saying you have no access to read Nextcloud groups. Then would it be possible for this option 3 to just create a **Calendar Distribution List - CDL** within Calendar app that can be maintained? Worst option, because everything already has these department/group lists in Nextcloud User Groups.

**Comment by jpubb** [comment]
> I'd offer the following:
> 1 - use access groups, especially synced LDAP groups, not groups from contacts. You can parse LDAP groups with standard LDAP tools.
**> 2 - if some participant account has no email, you can just add event to its calendar but do not send invitation by email.**

That's a good idea. I'd also include a notification for the user so they know it was added. But that's all extra code.  If I were the app author/maintainer, I'd just "no soup for you" if someone in the group doesn't have an email address for version 1 of this.  "Email adress for group member $username is missing. Stopping."

**Comment by jpubb** [comment]
> Nope, as always: This is a free-software project. Pull-requests are very much appreciated.

Btw, if anyone is interested in DIY instant bounties but you don't know how to code, you can always post the project for a fixed price at some place like Upwork, and make sending a pull request to the maintainer (in addition to sending the patch & instructions to you immediately) as part of the deliverable.

**Comment by msharp1** [comment]
Hi @tcitworld,
So, if I understand correctly you (framasoft?) don't aim to work on this feature anytime soon even with a bounty ?

**Comment by tcitworld** [comment]
I never committed to work on it, the bounty label was just dropped because [Nextcloud doesn't use Bountysource anymore](https://help.nextcloud.com/t/dont-use-bountysource-anymore/84943).

**Comment by mightynico** [comment]
> Will this function still be implemented?

+1

**Comment by julian70400** [comment]
Same here.. don't understand what's wrong with this feature.

**Comment by purejosh** [comment]
Still broken, somehow. Wow.

**Comment by q-wertz** [comment]
Issue is open, pull request still a draft. What did you expect?
And AFAIK it is a missing feature nothing broken.

**Comment by purejosh** [comment]
> And AFAIK it is a missing feature, nothing broken.
In the calendar app, the attendees box says "user group or email". I have no idea why you think this is a missing feature. Feature is literally touted in the application, and does not work.
